examples/QIR/Simulation/TraceSimulator/README.md

```diff
@@ -0,0 +1,281 @@
+- `IQuantumGateSet` : The Q# instruction set. Implementation of this interface is not strictly required, as long as *some* instruction set is implemented, and the QIR code only calls instructions from that set (may necessitate the use of a bridge, see the [QIR Bridge](https://github.com/microsoft/qsharp-runtime/tree/main/src/Qir/Runtime#qir-bridge-and-runtime) and the [top-level guide](../#understanding-the-qir-runtime-system)).
+- `IDiagnostics` : Optional interface to provide insight into the state of a simulator or hardware backend (useful for debugging).
+
+For a more detailed look at these interfaces, refer to the [top-level guide](../#structure-of-a-simulator) of the simulation example.
+
+To cleanly separate out different functionalities, the following file structure is used for the sample trace simulator:
+
+- `TraceSimulator.hpp` : Declaration of the simulator class, including required internal data structures and functions, as well as interface functions.
+- `QubitManager.hpp` : Simple qubit manager implementations to be used by the simulator.
+- `RuntimeManagement.cpp` : Implementation of all simulator functionality related to the `IRuntimeDriver` interface.
+- `TraceSimulation.cpp` : Implementation of all simulator functionality related to the `IQuantumGateSet` interface.
+
+## Trace Simulator Implementation
+
+The main components of each file are explained below, leaving out repetitive or boiler-plate code, the full extent of which can be viewed in the source files.
+
+---
+
+`TraceSimulator.hpp`
+
+As a basic trace simulator can be implemented in a state-less fashion, the `TraceSimulator` class' only member is a qubit manager instance.
+For consistency with the `StateSimulator`, we also use the same private functions to handle single-qubit and multi-controlled single-qubit gates:
+
+```cpp
+class TraceSimulator : public IRuntimeDriver, public IQuantumGateSet
+{
+    // Associated qubit manager instance to handle qubit representation.
+    QubitManager *qbm;
+
+    // To be called by quantum gate set operations.
+    void ApplyGate(Gate gate, Qubit target);
+    void ApplyControlledGate(Gate gate, long numControls, Qubit controls[], Qubit target);
+```
+
+For the trace simulator, the `Gate` type is represented by a string intended to hold the gate name:
+
+```cpp
+using Gate = std::string;
+```
+
+---
+
+`QubitManager.hpp`
+
+The sample qubit manager is kept simple, with the only required functionality besides allocation/deallocation being the ability convert a qubit to a printable string.
+In this case, the internal qubit ID representation is used:
+
+```cpp
+std::string GetQubitName(Qubit qubit)
+{
+    return std::to_string(GetQubitId(qubit));
+}
+```
+
+The qubit ID is a simple incrementing counter, and the `Qubit` object is just a pointer (as defined in [CoreTypes.hpp](https://github.com/microsoft/qsharp-runtime/blob/main/src/Qir/Runtime/public/CoreTypes.hpp)) with the qubit ID as its value:
```

Comment:
Actually, per my other comment, this is pobably a better place to explicitly call out that ids are never reused. From a tracing perspective, this can have an effect on total qubit count, which makes sense but might not be obvious to a user.